### ISSUE TITLE
ModelCheckpoint: saving initial_epoch, early stopping state, reduced lr, and training history

### ISSUE THREAD

**Issue by kl2792** [ecs-vlc/torchbearer#307]
So that we can resume training seamlessly from the epoch we KeyboardInterrupted at :-)

Happy to put in a PR for this. Any pointers on where to start?

**Comment by ethanwharris** [comment]
This is definitely something we’ve thought a lot about. At the moment, the call to Model.state_dict for saving only gets the model and the optimiser. However, we could have something that lets you choose more than that to be saved (e.g. callback states, for example). @MattPainter01 any thoughts?

**Comment by MattPainter01** [comment]
We'll probably have to talk more about it, but it should be easy to store more in the state_dict, we just need to be careful when to pull the values back out so that they are set properly in a fit call.

**Comment by kl2792** [comment]
Thoughts on saving the whole state dict instead of just model & optimizer and adding an argument to initialize training with a state dict? If pickling Callbacks is an issue, I can put in a __getstate__ method for them to contain just their parameters.

**Comment by ethanwharris** [comment]
@kl2792 We have been thinking about a good way to do this for a while. After some consideration we have decided to change the API by introducing a new class (which will eventually replace Model) that stores atomically, all necessary parameters for the fit calls (see #312). These classes will then implement the state_dict methods to persist only the required parameters so that they can be resumed (including a model history which auto populates the initial epoch) (this issue). Finally, we will add a mechanism for parameters to callbacks, fit calls etc. to be loaded from a seed state which enables grid search etc easily (see #314).

**Comment by kl2792** [comment]
That sounds a-mazing. The deep learning projects I did would have been heaven if I had a library like this one that had those features.

Do you think it's possible for the fit function to also print out previous epochs' tqdm bars based on the saved history?
